lib/ae_mdw/sync/server.ex

```diff
@@ -1,181 +1,353 @@
+  @typep state_data() :: %__MODULE__{
+           fork_height: height(),
+           db_state: State.t(),
+           mem_state: State.t(),
+           restarts: non_neg_integer(),
+           gens_per_min: non_neg_integer()
+         }
+  @typep cast_event() ::
+           {:new_height, height()}
+           | {:db_done, gens_per_min()}
+           | {:mem_done, gens_per_min()}
+           | {:fork, height()}
+           | :restart_sync
+  @typep internal_event() :: :check_sync
+  @typep call_event() :: :gens_per_min | :syncing?
+  @typep reason() :: term()
+  @typep info_event() :: {:DOWN, reference(), :process, pid(), reason()}
+
+  @max_restarts 5
   @retry_time 3_600_000
   @gens_per_min_weight 0.1
+  @mem_gens 1
+  @max_sync_gens 6
 
-  @type gens_per_min() :: number()
-  @opaque t() :: %__MODULE__{
-            syncing?: boolean(),
-            event_handler: EventHandler.t(),
-            gens_per_min: gens_per_min()
-          }
+  @spec start_link(GenStateMachine.options()) :: GenStateMachine.on_start()
+  def start_link(_opts), do:
+    GenStateMachine.start_link(__MODULE__, [], name: __MODULE__)
 
-  @spec start_link(GenServer.options()) :: GenServer.on_start()
-  def start_link(_opts),
-    do: GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  @spec new_height(height()) :: :ok
+  def new_height(height), do: GenStateMachine.cast(__MODULE__, {:new_height, height})
 
-  @spec start_sync() :: :ok
-  def start_sync, do: GenServer.cast(__MODULE__, :start_sync)
+  @spec fork(height()) :: :ok
+  def fork(height), do: GenStateMachine.cast(__MODULE__, {:fork, height})
 
-  @spec gens_per_min() :: number()
-  def gens_per_min, do: GenServer.call(__MODULE__, :gens_per_min)
+  @spec gens_per_min() :: gens_per_min()
+  def gens_per_min, do: GenStateMachine.call(__MODULE__, :gens_per_min)
 
   @spec syncing?() :: boolean()
-  def syncing?, do: GenServer.call(__MODULE__, :syncing?)
-
-  @spec done_db(State.t(), gens_per_min()) :: :ok
-  def done_db(state, gens_per_min),
-    do: GenServer.cast(__MODULE__, {:done_db, state, gens_per_min})
+  def syncing?, do: GenStateMachine.call(__MODULE__, :syncing?)
 
-  @spec done_mem(State.t(), gens_per_min()) :: :ok
-  def done_mem(state, gens_per_min),
-    do: GenServer.cast(__MODULE__, {:done_mem, state, gens_per_min})
+  @spec restart_sync() :: :ok
+  def restart_sync, do: GenStateMachine.cast(__MODULE__, :restart_sync)
 
   @impl true
+  @spec init([]) :: :gen_statem.init_result(state())
   def init([]) do
     db_store = DbStore.new()
     mem_store = MemStore.new(db_store)
 
-    db_state = State.new(db_store)
+    db_state = State.new()
     mem_state = State.new(mem_store)
 
-    event_handler = EventHandler.init(chain_height(), db_state, mem_state, &spawn_with_monitor/1)
+    state_data = %__MODULE__{
+      chain_height: nil,
+      fork_height: nil,
+      db_state: db_state,
+      mem_state: mem_state,
```

Comment:
Maybe there is a way to avoid exposing the whole database and in memory state to the State machine, having in the `state_data` only the state needed for the transitions.

Reply:
Same here

Reply:
The goal would be separate what is state needed to control the sync (the state machine `state_data`) from the state needed to execute the sync. But with the last refactor the use of `db_state` and `mem_state` are better encapsulated :+1: 

I would add only a comment explaining on the state machine notes that it needs only to compare the height of `db_state` and `mem_state` with the `chain_height` in order to decide what is the next transition (to the state `syncing_db` or `syncing_mem`).